cirq/linalg/transformations.py

```diff
@@ -304,3 +305,47 @@ def partial_trace(tensor: np.ndarray,
     left_indices = [keep_map[i] if i in keep_set else i for i in range(ndim)]
     right_indices = [ndim + i if i in keep_set else i for i in left_indices]
     return np.einsum(tensor, left_indices + right_indices)
+
+
+def keep_qubits(state: np.ndarray, keep_indices: List[int]) -> np.ndarray:
+    """Return a representation of the given state over a subset of its qubits.
+
+    A given wavefunction can be expressed over a subset of its qubits by
+    tracing out undesired qubits from the state's corresponding density matrix.
+    If the traced density matrix is rank 1 (pure state), then it can be
+    refactored into wavefunction form via eigendecomposition. Otherwise, the
+    outcome is a mixed state and cannot be expressed as a wavefunction.
+
+    Global phase of the input state will not necessariy be preserved in the
+    output state.
+
+    Args:
+        state: A wavefunction-like
+
+    """
+
+    if not np.isclose(np.linalg.norm(state), 1):
+        raise ValueError("Input state must be normalized.")
+    if not len(set(keep_indices)) == len(keep_indices):
+        raise ValueError(
+            "keep_indices were {} but must be unique.".format(keep_indices))
+    dim = np.log2(state.shape[0])
+    if not dim.is_integer():
+        raise ValueError("Input wavefunction has bad length of {}.".format(
+            state.shape[0]))
+    rho = np.kron(np.conj(np.reshape(state, (state.shape[0], 1)).T),
+                  state).reshape((2, 2) * int(dim))
+    new_dim = 2**len(keep_indices)
+    keep_rho = partial_trace(rho, keep_indices).reshape((new_dim,) * 2)
+    if not any(keep_indices):
+        return keep_rho
+
+    # FIXME: reasonable tolerance for purity check?
+    if approx_eq(np.trace(keep_rho @ keep_rho), 1):
+        w, v = np.linalg.eig(keep_rho)
+        # FIXME: reasonable tolerance for lone eigenvalue check?
```

Comment:
The success case of this method can be checked for significantly more efficiently than what is being done here. For each assignment of the other qubits, compute the 2-norm of the part of the array corresponding to that assignment, and then pick the one where the 2-norm was the largest. Normalize it to have a 2-norm of 1. This is the candidate wavefunction.

If the state factors into a "kept" and "not kept" part, the candidate wavefunction you just picked is equal to the kept part (modulo the fact that the global phase is not well defined). In order to verify that a candidate wavefunction is in fact a factor, you can dot product it against each other-assignment wavefunction and then add up the absolute values of the dot products. A result of 1 indicates perfect factoring.

In Quirk this is the strategy used by the amplitude display. The code is here: https://github.com/Strilanc/Quirk/blob/master/src/gates/AmplitudeDisplay.js#L65 . Unfortunately the code I linked is not easy to port over at all, because it is all tangled up with the way I go about executing it on the GPU.

If you create a circuit with an amplitude display and export the simulation data, it will include a "coherence measure" that is the value I was describing.

[![image](https://user-images.githubusercontent.com/79941/59138105-ea65dd00-893f-11e9-812c-faed5079e7dd.png)](https://algassert.com/quirk#circuit=%7B%22cols%22%3A%5B%5B%22H%22%5D%2C%5B%22%E2%80%A2%22%2C%22X%5E%E2%85%9B%22%5D%2C%5B%22Amps1%22%5D%5D%7D)


```
{
  [...]
 "displays": [
    [...]
    "serialized_id": "Amps1",
    "name": "Amplitude Display"
   },
   "data": {
    "coherence_measure": 0.9903004266906359,
    "superposition_phase_locked_state_index": 0,
    "ket": [
     {"r":0.7139325141906738,"i":0},
     {"r":0.6867600679397583,"i":0.13660506904125214}
    ],
    [...]
}
```

Reply:
Oh cool this looks much faster! I think I follow but want to double check - by "each assignment of the other qubits" and "part of the array corresponding to that assignment" do you mean some index-masking scheme that would take in (say) a 2-qubit wf [sqrt(.5), sqrt(.5), 0, 0] and compute the 1-qubit wf's [1, 0] (qubit 0) and [.5, .5] (qubit 1)? At first glance it seems this would yield the same candidate wf's given an input of |+++> or sqrt(.5)*(|000>+|111>) the same way (though Quirk handles the mixed state outcome in the GHZ case which is a good sign)

Do you know if the "coherence measure" being computed is equal to the "purity" (Tr(rho*rho)) for the factored state? This is seems a necessary condition for the above factoring scheme to be complete.

Reply:
I might be implementing your algorithm incorrectly; the way I did it it fails on the the example (|00> + |11>)|+> state I mentioned above. I've pushed scratch code of my implementation so maybe you can see an issue there?

Reply:
The coherence measure being computed is just some arbitrary thing that happens to equal 1 when pure; it doesn't correspond to any other measure AFAIK. Its only strength is that it is cheap to compute.

As an example of this not being a particularly ideal measure, it can vary based on operations performed on other qubits. For example, notice the text fading in and out as I move the QFT gate, corresponding to the measure going up and down:

![Peek 2019-06-17 17-43](https://user-images.githubusercontent.com/79941/59645595-9620e100-9127-11e9-9d49-bda12d2e439d.gif)

Reply:
> by "each assignment of the other qubits" and "part of the array corresponding to that assignment" do you mean some index-masking scheme that would take in (say) a 2-qubit wf [sqrt(.5), sqrt(.5), 0, 0] and compute the 1-qubit wf's [1, 0] (qubit 0) and [.5, .5] (qubit 1)?

I mean this:

```python
import cirq
import math
import numpy as np
def _candidate_sub_wavefunction(kept_qubit_indices, wavefunction):
    return max(
        (
            cirq.slice_for_qubits_equal_to(kept_qubit_indices, k)
            for k in range(len(wavefunction) >> len(kept_qubit_indices))
        ),
        key=lambda c: np.linalg.norm(c, 2))
```

Reply:
I mean this (I think; have not executed this code to test).

```python
import cirq
import math
import numpy as np
def _candidate_sub_wavefunction(kept_qubit_indices, wavefunction):
    # assumes wavefunction has shape (2,2,2,2,2,...)
    return max(
        (
            wavefunction[cirq.slice_for_qubits_equal_to(kept_qubit_indices, k)]
            for k in range(len(wavefunction) >> len(kept_qubit_indices))
        ),
        key=lambda c: np.linalg.norm(c, 2))
```

Reply:
Note: edited the code after noticing a mistake.